File: src/backend/decoders/ftp.py
def decodeFTP(rawPayload):
    """
    Decode FTP commands and responses from raw payload bytes.
    Returns a dict with Type (Command/Response), command/status, and argument/message,
    or None if the payload is not recognisable as FTP traffic.
    """
    FTP_COMMANDS = {
        "USER",
        "PASS",
        "ACCT",
        "CWD",
        "CDUP",
        "SMNT",
        "QUIT",
        "REIN",
        "PORT",
        "PASV",
        "TYPE",
        "STRU",
        "MODE",
        "RETR",
        "STOR",
        "STOU",
        "APPE",
        "ALLO",
        "REST",
        "RNFR",
        "RNTO",
        "ABOR",
        "DELE",
        "RMD",
        "MKD",
        "PWD",
        "LIST",
        "NLST",
        "SITE",
        "SYST",
        "STAT",
        "HELP",
        "NOOP",
        "FEAT",
        "OPTS",
        "MLST",
        "MLSD",
        "SIZE",
        "MDTM",
        "EPRT",
        "EPSV",
        "AUTH",
        "PBSZ",
        "PROT",
    }
    try:
        text = rawPayload.decode(errors="ignore")
        lines = text.replace("\r\n", "\n").split("\n")
        firstLine = lines[0].strip()
        if not firstLine:
            return None
        parts = firstLine.split(" ", 1)
        word = parts[0].upper()
        if word in FTP_COMMANDS:
            arg = parts[1].strip() if len(parts) > 1 else ""
            result = {
                "Type": "Command",
                "ftp.type": "Command",
                "Command": word,
                "ftp.command": word,
                "Argument": arg,
                "ftp.argument": arg,
            }
            if word == "USER" and arg:
                result["Credentials"] = {"username": arg}
            elif word == "PASS" and arg:
                result["Credentials"] = {"password": arg}
                result["Argument"] = "***"
                result["ftp.argument"] = "***"
            return result
        if len(word) == 3 and word.isdigit():
            statusCode = word
            message = parts[1].strip() if len(parts) > 1 else ""
            return {
                "Type": "Response",
                "ftp.type": "Response",
                "Status Code": statusCode,
                "ftp.status_code": statusCode,
                "Message": message,
                "ftp.message": message,
            }
        return None
    except Exception:
        return None

File: src/ui/decoders/conv/ftp.js
// FTP Conv decoder: parses "NNN " / "NNN-" response codes and recognized
// FTP commands (USER, PASS, LIST, RETR, STOR, PASV, PORT, etc.).

const FTP_COMMANDS = new Set([
    "USER",
    "PASS",
    "ACCT",
    "CWD",
    "CDUP",
    "PWD",
    "TYPE",
    "PASV",
    "EPSV",
    "PORT",
    "EPRT",
    "LIST",
    "NLST",
    "RETR",
    "STOR",
    "DELE",
    "RNFR",
    "RNTO",
    "MKD",
    "RMD",
    "SYST",
    "STAT",
    "FEAT",
    "AUTH",
    "QUIT",
    "NOOP",
]);

const FTP_RESPONSE_RE = /^(\d{3})([\s-])(.*)/;

function decodeFtpFromBytes(bytes) {
    const text = new TextDecoder("utf-8", { fatal: false }).decode(bytes);
    const lines = text.split(/\r?\n/).filter((line) => line.trim());
    if (!lines.length) return null;

    const fields = [];
    let detected = false;
    for (const line of lines) {
        const responseMatch = line.match(FTP_RESPONSE_RE);
        if (responseMatch) {
            const code = responseMatch[1];
            const suffix = responseMatch[2] === "-" ? " (cont.)" : "";
            fields.push({
                name: `Response ${code}${suffix}`,
                value: responseMatch[3] || "—",
            });
            detected = true;
        } else {
            const parts = line.trim().split(/\s+/);
            const command = (parts[0] || "").toUpperCase();
            if (FTP_COMMANDS.has(command)) {
                fields.push({ name: "Command", value: command });
                if (parts.length > 1) {
                    const argument = parts.slice(1).join(" ");
                    fields.push({
                        name: "Argument",
                        value: argument.length > 160 ? argument.slice(0, 160) + "…" : argument,
                    });
                }
                detected = true;
            }
        }
    }

    if (!detected) return null;
    return { protocol: "FTP", fields };
}

module.exports = { decodeFtpFromBytes };
